feat(chunkserver): Implement listeners for JobPool

dmga44 opened the pull request:
This PR targets changing the current behavior of the JobPool class,
in the way that it can serve different threads at the same time.
Current implementation does not allow a single JobPool instance to
serve different threads because it would mix up the wake up calls and
it would cause dangerous inconsistencies.

The proposed implementation partitions the JobPool inner structures for
each of the sources of jobs (which could be master or client network
workers) and creates specific channels of comunication to ensure that
only the source of a job is awaken.

This PR also includes changes in the JobPool's unittests to check the new
behavior.

The idea is to squash these commits.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
feat(cs): Implement listeners for JobPool

This change targets changing the current behavior of the JobPool class,
in the way that it can serve different threads at the same time.
Current implementation does not allow a single JobPool instance to
serve different threads because it would mix up the wake up calls and
it would cause dangerous inconsistencies.

The proposed implementation partitions the JobPool inner structures for
each of the sources of jobs (which could be master or client network
workers) and creates specific channels of comunication to ensure that
only the source of a job is awaken.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
refactor: Improve change quality

Changes:
- add comment when creating the JobPool instances to explain it a bit
better.
- add new/changed parameters to documentation.
- handle case when the provided vector for the wake up FDs is not the
expected size.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
test: Improve JobPool unittests

The multiplicity of sources of jobs of a single JobPool needs to be
tested, and this commit handles that.

Signed-off-by: Dave <dave@leil.io>

dmga44 commented on src/chunkserver/bgjobs.cc:
```diff
@@ -346,80 +379,86 @@ uint32_t job_replicate(JobPool &jobPool, JobPool::JobCallback callback, void *ex
-		                      processJob);
+		                      processJob, listenerId);
 	}
 	return job_invalid(jobPool, callback, extra);
```
Indeed, thank you for pointing that out.

dmga44 commented on src/chunkserver/bgjobs.cc:
```diff
@@ -346,80 +379,86 @@ uint32_t job_replicate(JobPool &jobPool, JobPool::JobCallback callback, void *ex
+		return jobPool.addJob(JobPool::ChunkOperation::ChangeVersion, callback, extra, processJob,
+		                      listenerId);
 	}
 	return job_invalid(jobPool, callback, extra);
```
Indeed, thank you for pointing that out.

dmga44 commented on src/chunkserver/masterconn.cc:
```diff
@@ -950,8 +950,12 @@ int masterconn_init_threads(void) {
+		// Create the JobPool instance with the specified number of workers, it would be serving
+		// only this master network thread, thus the number of listeners is 1.
+		std::vector<int> bgJobPoolFDs(1);
+		gJobPool = std::make_unique<JobPool>("ma", gNumberOfWorkers, kMaxBackgroundJobsCount, 1,
+		                                     bgJobPoolFDs);
+		gJobFD = bgJobPoolFDs[0];
```
You're right, thank you.

lgsilva3087 commented on src/chunkserver/bgjobs_unittest.cc:
```diff
@@ -62,45 +68,71 @@ class JobPoolTest : public ::testing::Test {
 	bool terminate;
 	std::mutex mutex_;
+	const static uint32_t kNrListeners = 4;
+	const static uint32_t kNrOperationTypes = 10;
```
These lines break building under my Clang, because the static variables are not defined outside the class. As they are compile time constants, please use:
```cpp
static constexpr uint32_t kNrListeners = 4;
static constexpr uint32_t kNrOperationTypes = 10;
```

dmga44 replied:
Applied.

lgsilva3087 commented on src/chunkserver/bgjobs.cc:
```diff
@@ -213,32 +241,30 @@ void JobPool::workerThread(const std::string &poolName, uint8_t workerId) {
-		        "JobPool: SendStatus: Failed to write to eventfd");
+void JobPool::sendStatus(uint32_t jobId, uint8_t status, uint32_t listenerId) {
+	auto &listenerInfo = listenerInfos_[listenerId];
+	std::unique_lock pipeLockGuard(listenerInfo.notifierMutex);
```
Please notice that there is no pipe anymore (`pipeLockGuard`).

dmga44 replied:
Indeed, I've renamed it to `statusLock`.

lgsilva3087 reviewed (commented):
So far looks great. Please see my comments.

lgsilva3087 reviewed (approved):
Great.

ralcolea commented on src/chunkserver/bgjobs.h:
```diff
@@ -166,20 +175,23 @@ class JobPool {
-		std::mutex jobsMutex;      /// Mutex for job operations.
-		std::unique_ptr<ProducerConsumerQueue> statusQueue;          /// Queue for job statuses.
+		int notifierFD;                                      /// File descriptor for notifications.
+		std::mutex notifierMutex;                            /// Mutex for pipe operations.
```
I think `notifierMutex` is not related with pipe operations.
Consider updating the documentation.

dmga44 replied:
Updated, now it says `/// Mutex for event notifications.`.

ralcolea commented on src/chunkserver/bgjobs.cc:
```diff
@@ -73,28 +86,35 @@ JobPool::~JobPool() {
 
 	jobsQueue.reset();
-	statusQueue.reset();
+	for (size_t i = 0; i < listenerInfos_.size(); ++i) { listenerInfos_[i].statusQueue.reset(); }
```
Consider using a range-based for loop as well as at the end of the function for closing the file descriptors.

dmga44 replied:
Applied, I used the analogous one to the end of the funcion for closing the file descriptors.

ralcolea commented on src/chunkserver/bgjobs.cc:
```diff
@@ -213,32 +241,32 @@ void JobPool::workerThread(const std::string &poolName, uint8_t workerId) {
+	if (listenerInfo.statusQueue->isEmpty()) {
+		static constexpr eventfd_t dummyValue = 1;  // Dummy value to wake up the eventfd
+		eassert(::eventfd_write(listenerInfo.notifierFD, dummyValue) == 0 &&
+		        "JobPool: SendStatus: Failed to write status to pipe");
```
I don't think this is a pipe but `eventfd` mechanism. It works similar to pipe, but it's more lightweight than pipes.
I think the previous message is correct, only the specific listener should be added to the message.

dmga44 replied:
Indeed, it was my mistake while rebasing some old WIP changes, thank you.

ralcolea commented on src/chunkserver/bgjobs_unittest.cc:
```diff
@@ -62,45 +68,71 @@ class JobPoolTest : public ::testing::Test {
-	// Wait for the job to be processed
-	std::this_thread::sleep_for(std::chrono::milliseconds(kWaitTimeMs));
+	for (int i = 0; i < nrJobsToProcess; ++i) {
+		auto opType = std::rand() % kNrOperationTypes;
```
The steps of selecting the operation and the target listener are not deterministic, which makes the unit test not deterministic and in case of failures, it makes more difficult to reproduce the test to identify the cause of the failure. I don't think this should remain in the final version.

I suggest you two different options:
- Use a deterministic random number generator with a fixed seed to make the test deterministic.
- As the number of jobs match the number of operation types, make sure to test always a different operation by assigning `opType = i`.

dmga44 replied:
I went with the first of your proposals:
- `std::srand(42);  // Seed for random number generation`

I don't like the second approach because it won't check the case when several jobs of the same type (or end up doing similar things) are being processed altogether, seemed too simple to me 😅 .

The idea of having it always random was to check with all the executions a broader set of scenarios (a new one every time), which will be fixed now. If the test were to have some probability `p` of failure, the expected time to reproduce (by modifying the test and logging the seed) is `1/p` seconds.

ralcolea commented on src/chunkserver/bgjobs.cc:
```diff
@@ -73,28 +86,35 @@ JobPool::~JobPool() {
-	std::unique_lock lock(jobsMutex);
-	uint32_t jobId = nextJobId++;
+                         ProcessJobCallback processJob, uint32_t listenerId) {
+	auto &listenerInfo = listenerInfos_[listenerId];
```
Many `JobPool` functions rely on having a valid `listenerId`.
Maybe, it's worth to validate `listenerId` is valid with an assertion, `std::exception` or any other mechanism to avoid possible runtime errors or crashes.

dmga44 replied:
I handled this case in most of the JobPool functions by logging the issue with `safs::log_warn` and directly returning, without doing the invalid operation.

For the `JobPool::addJob` function, besides the logging, I resetted the `listenerId` to 0 to make it valid and being able to perform the operation (an exception there could possibly crash the whole chunkserver).

ralcolea reviewed (commented):
Great job @dmga44!!!
Please, see my comments.

dmga44 committed:
refactor: Apply code review suggestions

Changes:
- handle the case when invalid listenerId is provided in most JobPool
functions.
- make deterministic the behavior of the JobPoolTest::ProcessJob
unittest.
- several documentation fixes and minor modernizations.

Co-authored-by: Crash <crash@leil.io>

Signed-off-by: Dave <dave@leil.io>

ralcolea reviewed (approved):
Great job @dmga44!!! :clap: :clap: :clap:
